resources/sdk/purecloudjava/tests/ApiClientRetryAsyncTest.java

```diff
@@ -0,0 +1,510 @@
+            stopwatch.stop();
+        }
+
+    }
+
+    @Test
+    public void invokeAsyncTestWith_504() throws IOException {
+        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
+
+        retryConfiguration.setMaxRetryTime(2);
+        retryConfiguration.setDefaultDelay(1000);
+
+        apiClient = getApiClient(retryConfiguration);
+
+        doReturn(getCloseableHttpResponse(504, Collections.emptyMap())).when(spyClient).execute(any(HttpUriRequest.class));
+
+        try {
+            stopwatch = Stopwatch.createStarted();
+            final Future<ApiResponse<ApiClientConnectorResponse>> response = apiClient.invokeAsync(getConnectorRequest(), getReturnType(), getAsyncApiCallBack());
+            final ApiResponse<ApiClientConnectorResponse> apiClientConnectorResponseApiResponse = response.get();
+        } catch (Exception e) {
+            Assert.assertEquals(504, ((ApiException) e.getCause()).getStatusCode());
+            Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 2000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 2100, "It will wait for every 1 sec and retry for 2 times");
+            stopwatch.stop();
+        }
+
+    }
+
+    @Test
+    public void invokeAsyncTestWith_504_And_No_MaxRetryTime() throws IOException {
+        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
+        apiClient = getApiClient(retryConfiguration);
+
+        doReturn(getCloseableHttpResponse(504, Collections.emptyMap())).when(spyClient).execute(any(HttpUriRequest.class));
+
+        try {
+            stopwatch = Stopwatch.createStarted();
+            final Future<ApiResponse<ApiClientConnectorResponse>> response = apiClient.invokeAsync(getConnectorRequest(), getReturnType(), getAsyncApiCallBack());
+            final ApiResponse<ApiClientConnectorResponse> apiClientConnectorResponseApiResponse = response.get();
+        } catch (Exception e) {
+            Assert.assertEquals(504, ((ApiException) e.getCause()).getStatusCode());
+            Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 0 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 100, "Since maxRetryTime is not provided it will not retry even if the status code is 504");
+            stopwatch.stop();
+        }
+
+    }
+
+    private AsyncApiCallback<ApiResponse<ApiClientConnectorResponse>> getAsyncApiCallBack() {
+        return new AsyncApiCallback<ApiResponse<ApiClientConnectorResponse>>() {
+            @Override
+            public void onCompleted(ApiResponse<ApiClientConnectorResponse> response) {
+            }
+
+            @Override
+            public void onFailed(Throwable exception) {
+            }
+        };
+    }
+
+    private CloseableHttpResponse getCloseableHttpResponse(int statusCode, Map<String, String> headerMap) {
```

Comment:
Can this be a mock instead of a concrete impl?

Reply:
Could be but I wanted to use different status codes and headers for different tests so making it generic.
couldn't help myself for implementing all the unwanted methods though :(

Reply:
I think I can not. once I create a mocked object: 
`CloseableHttpResponse mockResponse = Mockito.mock(CloseableHttpResponse.class);`
I can't set anything as being mocked one but I want to set `statuCode` and `header` dynamically for each test.

Reply:
@deerghayu this still needs to be addressed

Reply:
`CloseableHttpResponse mockResponse = Mockito.mock(CloseableHttpResponse.class);
        mockResponse.setStatusLine(new ProtocolVersion("HTTP/1.1", 1, 1), 429, "Bad GateWay");
        mockResponse.setHeader("Retry-after", "1");`

` doReturn(mockResponse).when(spyClient).execute(any(HttpUriRequest.class));`

If you were suggesting something like this, it won't work since status code and header will not be set as being a mock object. (I'm trying to set different status code and header for each test)

Please correct me if I'm missing something from the comment.

Reply:
I refactored the mock Response based on the suggestion. Thanks for the comment.